### PR TITLE
Feat(eos_designs): Add support for ipv4_acls with field replacement

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

Add support for ipv4_acls with field replacement.
Currently only applicable under node l3_interfaces.

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes

```yaml
# IPv4 Extended Access-lists supporting substition on certain fields.
# The substition is useful when assigning the same access-list on multiple interfaces,
# but where certain fields require unique values like the "interface_ip" or "peer_ip".
# When using substition, the interface name will be appended to the ACL name.
ipv4_acls:

    # Access-list name.
    # When using substition for any fields, the interface name will be appended to the ACL name.
  - name: <str; required; unique>

    # ACL Entries.
    entries: # required

        # This field supports substition of the fields "interface_ip" and "peer_ip".
        # Alternatively it can be set with a static value of "any", "<ip>/<mask>" or "<ip>".
        # "<ip>" without a mask means host.
        # Required except for remarks.
      - source: <str>

        # This field supports substition of the fields "interface_ip" and "peer_ip".
        # Alternatively it can be set with a static value of "any", "<ip>/<mask>" or "<ip>".
        # "<ip>" without a mask means host.
        # Required except for remarks.
        destination: <str>

      ...remaining ip_access_list options from eos_cli_config_gen...
```

## How to test

Added molecule for positive test case. Manually tested various negative test cases.

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/ipv4_acls.schema.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+# yaml-language-server: $schema=../../../../plugins/plugin_utils/schema/avd_meta_schema.json
+# Line above is used by RedHat's YAML Schema vscode extension
+# Use Ctrl + Space to get suggestions for every field. Autocomplete will pop up after typing 2 letters.
+type: dict
+keys:
+  ipv4_acls:
+    type: list
+    primary_key: name
+    description: |-
+      IPv4 Extended Access-lists supporting substition on certain fields.
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/ipv4_acls.schema.yml]:
```suggestion
      IPv4 Extended Access-lists supporting substitution on certain fields.
```
Same in other places

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/underlay/utils.py [hunk 2]
```diff
@@ -279,3 +281,44 @@ def _get_l2_as_subint(self, link: dict, svi: dict, vrf: dict) -> dict:
...
+            if interface_ip is not None and "/" in interface_ip:
+                interface_ip = interface_ip.split("/", maxsplit=1)[0]
+            peer_ip: str | None = get(l3_interface, "peer_ip")
+
+            if ipv4_acl_in is not None:
+                l3_interface_acls.setdefault(interface_name, {})["ipv4_acl_in"] = self.shared_utils.get_ipv4_acl(
+                    name=ipv4_acl_in,
+                    interface_name=interface_name,
+                    interface_ip=interface_ip,
+                    peer_ip=peer_ip,
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/underlay/utils.py:329]: we may need to refactor later with some data class if we start supporting lots of params

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_access_lists.schema.yml [hunk 3]
```diff
@@ -10,84 +10,96 @@ keys:
...
-                description: |
-                  ACL entry sequence number.
+                  - str
+                description: ACL entry sequence number.
               remark:
                 type: str
-                description: |
+                description: |-
                   Comment up to 100 characters.
                   If remark is defined, other keys in acl entry will be ignored.
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_access_lists.schema.yml]: While we are fixing the schema..
```suggestion
                  If remark is defined, other keys in ACL entry will be ignored.
```

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/misc.py [hunk 4]
```diff
@@ -307,3 +309,58 @@ def new_network_services_bgp_vrf_config(self: SharedUtils) -> bool:
...
         return get(self.hostvars, "new_network_services_bgp_vrf_config", default=default_value)
+
+    @cached_property
+    def ipv4_acls(self: SharedUtils) -> list:
+        return get(self.hostvars, "ipv4_acls", default=[])
+
+    def get_ipv4_acl(self: SharedUtils, name: str, interface_name: str, *, interface_ip: str | None = None, peer_ip: str | None = None):
+        """
+        Get one IPv4 ACL from "ipv4_acls" where fields have been substituted.
+        If any substition is done, the ACL name will get "_<interface_name>" appended.
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/misc.py]:
```suggestion
        If any substitution is done, the ACL name will get "_<interface_name>" appended.
```

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/structured_configs/ipv4-acls.yml [hunk 5]
```diff
@@ -0,0 +1,157 @@
+hostname: ipv4-acls
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/structured_configs/ipv4-acls.yml:1]: do you want to sneak in the gitattribute thing for structured_config we talked about?

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/ipv4-acls.cfg [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ip access-list TEST-IPV4-ACL-WITH-IP-FIELDS-OUT_Ethernet1
+   remark Some remark will not require source and destination fields.
+   permit ip host 172.16.0.2 any
+!
+ip access-list TEST-IPV4-ACL-WITH-IP-FIELDS-OUT_Ethernet2
+   remark Some remark will not require source and destination fields.
+   permit ip host 172.17.17.17 any
+!
+ip access-list TEST-IPV4-ACL-WITH-NO-FIELDS-IN
+   permit ip host host 172.18.18.18 any
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/ipv4-acls.cfg]: duplicate host?

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/host_vars/ipv4-acls.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    entries:
+      - action: permit
+        protocol: ip
+        source: host 172.18.18.18
+        destination: any
+  - name: TEST-IPV4-ACL-WITH-NO-FIELDS-OUT
+    entries:
+      - action: permit
+        protocol: ip
+        source: host 172.18.18.18
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/host_vars/ipv4-acls.yml]:
```suggestion
        source: 172.18.18.18
```

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/misc.py [hunk 8]
```diff
@@ -307,3 +309,58 @@ def new_network_services_bgp_vrf_config(self: SharedUtils) -> bool:
...
+          value: replacement value to set
+
+        If a replacement is done, but the value is None, an error will be raised.
+        """
+        for key, value in replacements.items():
+            if field_value != key:
+                continue
+
+            if value is None:
+                raise AristaAvdError(
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/misc.py:358]: negative unit test for this one?

**Review by gmuloc — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/structured_configs/ipv4-acls.yml (hunk 5 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/structured_configs/ipv4-acls.yml:1]: nope

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/host_vars/ipv4-acls.yml (hunk 7 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/host_vars/ipv4-acls.yml]: fixed

File: ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/ipv4_acls.schema.yml (hunk 1 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/ipv4_acls.schema.yml]: Not sure why my brain has made the same spelling mistake all over the place...

File: ansible_collections/arista/avd/roles/eos_designs/docs/tables/wan-interfaces-settings.md [hunk 9]
```diff
@@ -164,9 +164,10 @@
             ip_address: <str>
 
             # When the `ip_address` is `dhcp`, this optional field allows to indicate the expected
-            # IPv4 address/Mask to be allocated on the interface if known.
-            # This is not rendered in the configuration but can be used in ACLs in certain scenarios.
-            dhcp_ip_address: <str>
+            # IPv4 address (without mask) to be allocated on the interface if known.
+            # This is not rendered in the configuration but can be used for substution of 'interface_ip' in the Access-list
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/docs/tables/wan-interfaces-settings.md]:
```suggestion
            # This is not rendered in the configuration but can be used for substitution of 'interface_ip' in the Access-list
```

**Review by gmuloc — APPROVED** [review]
LGTM

File: ansible_collections/arista/avd/roles/eos_designs/docs/tables/ipv4-acls.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    | [<samp>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;dscp</samp>](## "ipv4_acls.[].entries.[].dscp") | String |  |  |  | DSCP value or name. |
+    | [<samp>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;vlan_number</samp>](## "ipv4_acls.[].entries.[].vlan_number") | Integer |  |  |  |  |
+    | [<samp>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;vlan_inner</samp>](## "ipv4_acls.[].entries.[].vlan_inner") | Boolean |  | `False` |  |  |
+    | [<samp>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;vlan_mask</samp>](## "ipv4_acls.[].entries.[].vlan_mask") | String |  |  |  | 0x000-0xFFF VLAN mask. |
+    | [<samp>&nbsp;&nbsp;&nbsp;&nbsp;counters_per_entry</samp>](## "ipv4_acls.[].counters_per_entry") | Boolean |  |  |  |  |
+
+=== "YAML"
+
+    ```yaml
+    # IPv4 Extended Access-lists supporting substitution on certain fields.
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/docs/tables/ipv4-acls.md]: Add details about where to use these ACLs.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/underlay/utils.py [hunk 11]
```diff
@@ -185,6 +185,8 @@ def _get_l3_interface_cfg(self, l3_interface: dict) -> dict | None:
             "description": interface_description,
             "speed": l3_interface.get("speed"),
             "service_profile": l3_interface.get("qos_profile"),
+            "access_group_in": get(self._l3_interface_acls, f"{interface_name}.ipv4_acl_in.name"),
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/underlay/utils.py:188]: Add some check and raise if there is no ipv4_acl_in set for wan interface.
Also add to doc.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/underlay/utils.py [hunk 12]
```diff
@@ -198,6 +200,12 @@ def _get_l3_interface_cfg(self, l3_interface: dict) -> dict | None:
         if self.shared_utils.is_cv_pathfinder_router:
             interface["flow_tracker"] = {"hardware": self.shared_utils.wan_flow_tracker_name}
 
+        if self.shared_utils.is_wan_router and (wan_carrier_name := l3_interface.get("wan_carrier")) is not None and interface["access_group_in"] is None:
+            if not get(get_item(self.shared_utils.wan_carriers, "name", wan_carrier_name, default={}), "trusted"):
+                raise AristaAvdError(
+                    f"'ipv4_acl_in' must be set on WAN interfaces where 'wan_carrier' is set. 'ipv4_acl_in' is missing on interface '{interface_name}'"
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/underlay/utils.py]: do we want to add sthing about "trusted" there?

File: ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/defs_node_type_l3_interfaces.schema.yml [hunk 13]
```diff
@@ -76,6 +77,21 @@ $defs:
           type: str
           description: |-
             The peer device IPv4 address (no mask). Used as default route gateway if `set_default_route` is true and `ip` is an IP address.
+        ipv4_acl_in:
+          description: |-
+            Name of the IPv4 Access-list to be assigned in the ingress direction.
+            The access-list must be defined under `ipv4_acls`.
+            Required for all WAN interfaces (`wan_carrier` is set) unless the Carrier is marked as 'trusted' under `wan_carriers`.
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/defs_node_type_l3_interfaces.schema.yml]:
```suggestion
            Required for all WAN interfaces (`wan_carrier` is set) unless the carrier is marked as 'trusted' under `wan_carriers`.
```

File: ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/defs_node_type_l3_interfaces.schema.yml [hunk 14]
```diff
@@ -95,6 +111,7 @@ $defs:
           description: |-
             The WAN Carrier this interface is connected to.
             This is used to infer the path-groups in which this interface should be configured.
+            Unless the Carrier is marked as 'trusted' under `wan_carriers`, `ipv4_acl_in` is also required on all WAN interfaces.
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/defs_node_type_l3_interfaces.schema.yml]: Let's change`WAN Carrier` to `WAN carrier` .. I am upper-casing too much sometimes :)
```suggestion
            Unless the carrier is marked as 'trusted' under `wan_carriers`, `ipv4_acl_in` is also required on all WAN interfaces.
```

File: ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/wan_carriers.schema.yml [hunk 15]
```diff
@@ -26,3 +26,8 @@ keys:
           type: str
           description: The path-group to which this carrier belongs.
           required: true
+        trusted:
+          type: bool
+          default: false
+          description: |-
+            WAN interfaces will require an inbound Access-list to be set, unless the provider is "trusted".
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/schemas/schema_fragments/wan_carriers.schema.yml]:
```suggestion
            Set this to `true` to mark this carrier as "trusted".
            WAN interfaces require an inbound Access-list to be set unless the carrier is "trusted".
```

**Review by gmuloc — CHANGES_REQUESTED** [review]

**Review by gmuloc — APPROVED** [review]

**Review by carlbuchmann — APPROVED** [review]
LGTM tested in lab topology, works well and error messages are useful!

**Pull request merged**